# Task 1: Permutations and Combinations

> Suppose we alter the Lady Tasting Tea experiment to involve twelve cups of tea. Six have the milk in first and the other six having tea in first. A person claims they have the special power of being able to tell whether the tea or the milk went into a cup first upon tasting it. You agree to accept their claim if they can tell which of the six cups in your experiment had the milk in first.

> Calculate, using Python, the probability that they select the correct six cups. Here you should assume that they have no special powers in figuring it out, that they are just guessing. Remember to show and justify your workings in code and MarkDown cells.

> Suppose, now, you are willing to accept one error. Once they select the six cups they think had the milk in first, you will give them the benefit of the doubt should they have selected at least five of the correct cups. Calculate the probability, assuming they have no special powers, that the person makes at most one error.

> Would you accept two errors? Explain.

We will start by giving the coups different numbers, the numbers will not have any influence on the cups order

1 2 3 4 5 6 7 8 9 10 11 12    

In [45]:

# number of cups in total
no_cups = 12

# number of cups with milk first 
no_cups_milk_first = 6

# number of cups with tea first
no_cups_tea_first = 6

> How many different ways are there of selecting 4 things of 8?

In [46]:
# Mathematical functions from the standard library
# https://docs.python.org/3/library/math.html
import math

# Permutations and combinations
# https://docs.python.org/3/library/itertools.html
import itertools


# Random selections
# https://docs.python.org/3/library/random.html#module-random
import random

In [47]:
# number of ways of selecting four cups from eight
ways = math.comb(no_cups, no_cups_milk_first) 

# show one of the combinations, i.e.: x x 3 4 5 x x 8 x x 11 12
ways 

924

1 2 3 4 5 6 7 8 9 10 11 12

: _ _ _ _ _ _

In [48]:
# total number of ways of selecting four cups from 12, keeping them in order 
# for the first slot, we have 12 options, for the second it will be 11, for the third 10, and the fourth 9 

ways_ordered = 12 * 11 * 10 * 9 * 8 * 7

ways_ordered

665280

There are 665280 ways of selecting 6 cups out of 12

Is the order important? Considering that we pour milk first in the following cups:  
3 5 7 8 10 11  
7 8 5 3 11 10    

> The answer is no, the important is that they all contain milk first

In [49]:
# no of ways of shuffling the four cups 
no_shuffles = 6 * 5 * 4 * 3 * 2 * 1

no_shuffles

720

There are 720 ways to shuffle 6 cups

In [50]:
# number of combinations 
no_combs = ways_ordered // no_shuffles
no_combs

924

> Repeating the same calculation using math.factorial()

In [51]:
n = 12
k = 6

# Twelve factorial 
math.factorial(n)

479001600

In [52]:
# Six factorial
math.factorial(k)

720

In [53]:
math.factorial(n) // (math.factorial(k) * math.factorial(n - k))

924

# Ordering of results

https://docs.python.org/3/library/itertools.html#itertools.combinations

In [54]:
# Cup labels
labels = list(range(no_cups))

# Show
labels

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [55]:
# show the different ways of selecting no_cups_milk_first out of no_cups cups of tea
combs = list (itertools.combinations(labels, no_cups_milk_first))

combs


[(0, 1, 2, 3, 4, 5),
 (0, 1, 2, 3, 4, 6),
 (0, 1, 2, 3, 4, 7),
 (0, 1, 2, 3, 4, 8),
 (0, 1, 2, 3, 4, 9),
 (0, 1, 2, 3, 4, 10),
 (0, 1, 2, 3, 4, 11),
 (0, 1, 2, 3, 5, 6),
 (0, 1, 2, 3, 5, 7),
 (0, 1, 2, 3, 5, 8),
 (0, 1, 2, 3, 5, 9),
 (0, 1, 2, 3, 5, 10),
 (0, 1, 2, 3, 5, 11),
 (0, 1, 2, 3, 6, 7),
 (0, 1, 2, 3, 6, 8),
 (0, 1, 2, 3, 6, 9),
 (0, 1, 2, 3, 6, 10),
 (0, 1, 2, 3, 6, 11),
 (0, 1, 2, 3, 7, 8),
 (0, 1, 2, 3, 7, 9),
 (0, 1, 2, 3, 7, 10),
 (0, 1, 2, 3, 7, 11),
 (0, 1, 2, 3, 8, 9),
 (0, 1, 2, 3, 8, 10),
 (0, 1, 2, 3, 8, 11),
 (0, 1, 2, 3, 9, 10),
 (0, 1, 2, 3, 9, 11),
 (0, 1, 2, 3, 10, 11),
 (0, 1, 2, 4, 5, 6),
 (0, 1, 2, 4, 5, 7),
 (0, 1, 2, 4, 5, 8),
 (0, 1, 2, 4, 5, 9),
 (0, 1, 2, 4, 5, 10),
 (0, 1, 2, 4, 5, 11),
 (0, 1, 2, 4, 6, 7),
 (0, 1, 2, 4, 6, 8),
 (0, 1, 2, 4, 6, 9),
 (0, 1, 2, 4, 6, 10),
 (0, 1, 2, 4, 6, 11),
 (0, 1, 2, 4, 7, 8),
 (0, 1, 2, 4, 7, 9),
 (0, 1, 2, 4, 7, 10),
 (0, 1, 2, 4, 7, 11),
 (0, 1, 2, 4, 8, 9),
 (0, 1, 2, 4, 8, 10),
 (0, 1, 2, 4, 8, 11),
 (0, 1, 2, 4

In [56]:
# Number of combinations
len(combs)

924

In [72]:
# Select six cups at random to put milk in first
# https://docs.python.org/3/library/random.html#random.sample
labels_milk = random.sample(labels, 6)

# sort, inplace
labels_milk.sort()

# show
labels_milk

[4, 6, 7, 9, 10, 11]

______

# Task 2: numpy's Normal Distribution

> In this task you will assess whether numpy.random.normal() properly generates normal values. To begin, generate a sample of one hundred thousand values using the function with mean 10.0 and standard deviation 3.0.

> Use the scipy.stats.shapiro() function to test whether your sample came from a normal distribution. Explain the results and output.

> Plot a histogram of your values and plot the corresponding normal distribution probability density function on top of it.

